rfcs/lower-bounds-calculation.md

```diff
@@ -0,0 +1,215 @@
+
+```lua
+-- D.lua
+local part = script.Parent
+
+local function blink()
+    -- ...
+end
+
+part.Touched:Connect(blink)
+```
+
+The `Touched` event actually passes a single argument: the part that touched the `Instance` in question.  In this example, it is omitted from the callback handler.
+
+We therefore want _oversaturation_ of a function to be allowed, but this combines with optional function arguments to create a problem with soundness.  Consider the following:
+
+```lua
+-- E.lua
+type Callback = (Instance) -> ()
+
+local cb: Callback
+function register_callback(c: Callback)
+    cb = c
+end
+
+function invoke_callback(i: Instance)
+    cb(i, e)
+end
+
+---
+
+function bad_callback(x: number?)
+end
+
+local obscured: () -> () = bad_callback
+
+register_callback(obscured)
+
+function good_callback()
+end
+
+register_callback(good_callback)
+```
+
+The problem we run into is, if we allow the subtyping rule `(T?) -> () <: () -> ()` and also allow oversaturation of a function, it becomes easy to obscure an argument type and pass the wrong type of value to it.
+
+Next, consider the following type alias
+
+```lua
+-- F.lua
+type OldFunctionType = (any, any) -> any
+type NewFunctionType = (any) -> any
+type FunctionType = OldFunctionType & NewFunctionType
+```
+
+If we have a subtyping rule `(T0..TN) <: (T0..TN-1)` to permit the function subtyping relationship `(T0..TN-1) -> R <: (T0..TN) -> R`, then the above type alias normalizes to `(any) -> any`.  In order to call the two-argument variation, we would need to permit oversaturation, which runs afoul of the soundness hole from the previous example.
+
+We need a solution here.
+
+To resolve this, let's reframe things in simpler terms:
+
+If there is never a subtyping relationship between packs of different length, then we don't have any soundness issues, but we find ourselves unable to register `good_callback`.
+
+To resolve _that_, consider that we are in truth being a bit hasty when we say `good_callback : () -> ()`.  We can pass any number of arguments to this function safely.  We could choose to type `good_callback : () -> () & (any) -> () & (any, any) -> () & ...`.  Luau already has syntax for this particular sort of infinite intersection: `good_callback : (any...) -> ()`.
+
+So, we propose some different inference rules for functions:
+
+1. The AST fragment `function(arg0..argN) ... end` is typed `(T0..TN, any...) -> R` where `arg0..argN : T0..TN` and `R` is the inferred return type of the function body.  Function statements are inferred the same way.
+1. Type annotations are unchanged.  `() -> ()` is still a nullary function.
+
+And subtyping rules for type packs:
+
+1. `(T0..TN) <: (U0..UN)` if, for each `T` and `U`, `T <: U`
+
+1. `A & B <: T` if `T <: A` or `T <: B` is the standard left-side subtyping rule for intersections.  It is unchanged.  We include it here for clarity.
+1. `(U...) ~ () & (U) & (U, U) & (U, U, U) & ...`, therefore
```

Comment:
I think here you want unions?
The idea being to think of `(T...) -> U` as `(() -> U) & ((T) -> U) & ((T,T) -> U) & ...` and since function arguments are contravariant, this is the same as `(() | (T) | (T,T) | ...) -> U`.